File: src/api/consumer/finance.rs
use axum::routing::{get, post};
use crate::api::web_path::{WebPath, WebPathType};
use crate::application::consumer::finance_service;
use crate::common::error::Error;
use crate::common::result::ApiResponse;
use crate::model::prelude::ListData;
use crate::domain::args::a_finance::*;
use crate::domain::model::m_finance::*;
use axum::{extract::{Path, Query}, Json};
use rust_decimal::Decimal;
use validator::Validate;

pub async fn get_account(
    Path(consumer_id): Path<i64>,
) -> Result<Json<ApiResponse<AccountInfoResp>>, Error> {
    let result = finance_service::get_account(consumer_id).await?;

    Ok(Json(ApiResponse::success(result)))
}

pub async fn recharge(
    Path(consumer_id): Path<i64>,
    Json(args): Json<RechargeArgs>,
) -> Result<Json<ApiResponse<TransactionModel>>, Error> {
    args.validate()?;

    let amount = Decimal::try_from(args.amount * 100.0).unwrap_or(Decimal::ZERO);

    let result = finance_service::recharge(RechargeParams {
        consumer_id,
        amount,
        related_order_no: Some(args.payment_order_no),
    })
    .await?;

    Ok(Json(ApiResponse::success(result)))
}

pub async fn withdraw(
    Path(consumer_id): Path<i64>,
    Json(args): Json<WithdrawArgs>,
) -> Result<Json<ApiResponse<TransactionModel>>, Error> {
    args.validate()?;

    let amount = Decimal::try_from(args.amount * 100.0).unwrap_or(Decimal::ZERO);

    let result = finance_service::withdraw(WithdrawParams {
        consumer_id,
        amount,
    })
    .await?;

    Ok(Json(ApiResponse::success(result)))
}

pub async fn list_transactions(
    Query(params): Query<TransactionListArgs>,
) -> Result<Json<ApiResponse<ListData<TransactionModel>>>, Error> {
    let page_size = params.page_size;
    let page_num = params.page_num;
    let (items, total) = finance_service::list_transactions(
        TransactionListParams {
            page_num: params.page_num,
            page_size: params.page_size,
            transaction_type: params.transaction_type,
            start_time: params.start_time,
            end_time: params.end_time,
            consumer_id: params.consumer_id,
        },
    )
    .await?;

    Ok(Json(ApiResponse::success(ListData {
        list: items,
        total,
        total_pages: (total + page_size.unwrap_or(10) as u64 - 1) / page_size.unwrap_or(10) as u64,
        page_num: page_num.unwrap_or(1) as u64,
    })))
}

pub fn finance_api() -> WebPath {
    WebPath::new()
        .route("/account/{id}", WebPathType::Get, Some("获取账户信息"), get(get_account))
        .route("/recharge/{id}", WebPathType::Post, Some("充值"), post(recharge))
        .route("/withdraw/{id}", WebPathType::Post, Some("提现"), post(withdraw))
        .route("/transactions", WebPathType::Get, Some("交易流水列表"), get(list_transactions))
}

File: frontend/src/api/modules/consumer/finance.ts
import axios from "axios";
import { Message } from "@arco-design/web-vue";

export interface AccountInfo {
  id: number;
  consumer_id: number;
  balance: string;
  frozen_balance: string;
  available_balance: string;
  created_at?: string;
  updated_at?: string;
}

export interface Transaction {
  id: number;
  consumer_id: number;
  transaction_no: string;
  transaction_type: "recharge" | "consume" | "withdraw" | "refund";
  amount: string;
  balance_before: string;
  balance_after: string;
  description?: string;
  related_order_no?: string;
  operator_id?: number;
  created_at?: string;
}

export interface RechargeParams {
  amount: number;
  payment_order_no: string;
}

export interface WithdrawParams {
  amount: number;
}

export interface PageParams {
  page_num?: number;
  page_size?: number;
  transaction_type?: string;
  start_time?: string;
  end_time?: string;
}

export const financeApi = {
  getAccount: async (consumer_id: number) => {
    const res = await axios.get(`/finance/account/${consumer_id}`);
    return res.data;
  },

  recharge: async (consumer_id: number, params: RechargeParams) => {
    const res = await axios.post(`/finance/recharge/${consumer_id}`, params);
    if (res.data.message !== "success") {
      Message.error(res.data.message);
    }
    return res.data;
  },

  withdraw: async (consumer_id: number, params: WithdrawParams) => {
    const res = await axios.post(`/finance/withdraw/${consumer_id}`, params);
    if (res.data.message !== "success") {
      Message.error(res.data.message);
    }
    return res.data;
  },

  transactions: async (params: PageParams) => {
    const res = await axios.get("/finance/transactions", { params });
    return res.data;
  },

  statistics: async (consumer_id?: number) => {
    const res = await axios.get("/finance/statistics", { params: { consumer_id } });
    return res.data;
  }
};
